### ISSUE TITLE
youku 201

### ISSUE THREAD

**Issue by BenjaminGao** [soimort/you-get#2056]
![screen shot 2017-06-01 at 9 12 19 pm](https://cloud.githubusercontent.com/assets/16028320/26681316/5ef8c67a-470f-11e7-9a1e-ce6591429213.png)

```
➜  ~ you-get -d http://v.youku.com/v_show/id_XMjc5NzcyNzU4NA\=\=.html\?from\=s1.8-1-1.1\&spm\=a2h0k.8191407.0.0
[DEBUG] get_content: http://play.youku.com/play/get.json?vid=XMjc5NzcyNzU4NA==&ct=10
[DEBUG] get_content: http://play.youku.com/play/get.json?vid=XMjc5NzcyNzU4NA==&ct=12
you-get: [Failed] 客户端无权播放,201
```

**Comment by alanyang** [comment]
ups接口也不能用了

**Comment by rosynirvana** [comment]
@alanyang 
多试几次……

**Comment by lamlam123tld3** [comment]
ups requires param utid now
which can be retrieve from the response cookie cna from http://gm.mmstat.com/yt/ykcomment.play.commentInit

one awareness is if the cookie is loaded,
if cna is in the cookie
use this cna instead of fetch a new one

**Comment by pxia** [comment]
```
$ /Users/pxia/work/you-get/you-get -c ../cookies.txt http://v.youku.com/v_show/id_XMTgxMjEwMDE0OA\=\=.html
you-get: Found cna in imported cookies. Use it
you-get: 客户端无权播放,201
```

Loading a VIP cookie with cna also gives me 201.

**Comment by rosynirvana** [comment]
Does exported cookies guarantee a un-urlencoded cookie value?

**Comment by pxia** [comment]
My bad, there was a plus sign in the cna from cookies. I think you need to quote that as well in your pull request.

**Comment by pxia** [comment]
It depends on the specification of that file format. (e.g. sqlite/txt)
But how the data is encoded in the file doesn't matter, the data saved in memory should always be decoded. It's handled by `http.cookiejar` when reading from the file.

**Comment by rosynirvana** [comment]
Is it a documented feature of ```http.cookiejar```?

**Comment by pxia** [comment]
I don't know, but I'm very certain this is the case.

**Comment by rosynirvana** [comment]
I'm sure ```http.cookiejar``` won't unquote some cookie values.
```python
import http.cookiejar as jar

cookies = jar.MozillaCookieJar('ck.txt')
cookies.load()

for cookie in cookies:
    if '%' in cookie.value:
        print(cookie.value)
```

```
$ python3 unquote.py
150440357.1485531158.1.1.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=(not%20provided)
72494102.1486221075.1.1.utmcsr=google|utmccn=(organic)|utmcmd=organic|utmctr=(not%20provided)
1699347926-1483604304-http%253A%252F%252Fg.e-hentai.org%252F%7C1483604304
%5B%22%22%2C%22%22%2C1486881607%2C%22https%3A%2F%2Fwww.google.com%2F%22%5D
739118017-1482660326-null%7C1485879215
cnzz_eid%3D1271487756-1484671251-null%26ntime%3D1484671251
1929808911-1486881650-https%253A%252F%252Fwww.google.com%252F%7C1486881650
cnzz_eid%3D279007051-1486181763-null%26ntime%3D1486181763
cnzz_eid%3D282595278-1486474390-null%26ntime%3D1486474390
```

**Comment by alkorsan** [comment]
@rosynirvana 
Hi, 
your patch was working, but yesterday it stopped working, I did update with your new update but I get the same error:

`you-get: 客户端无权播放,201`

thank you very much for your help

**Comment by rosynirvana** [comment]
by 
> update with your new update
you mean https://github.com/soimort/you-get/pull/2015/commits/a0616b1829ff9c7feb214ab32deb98e9ffab1614 ?

**Comment by alkorsan** [comment]
@rosynirvana 
yes , I mean a0616b1
thanks

**Comment by rosynirvana** [comment]
Then the debug output?

**Comment by alkorsan** [comment]
@rosynirvana

[DEBUG] get_content: https://ups.youku.com/ups/get.json?vid=XMjYxMjEyNDU0MA==&cc
ode=0401&client_ip=192.168.1.1&utid=6bf31135&client_ts=1496418780
you-get: 客户端无权播放,201

thanks

**Comment by rosynirvana** [comment]
And it can be reproduced every time for you?

**Comment by rosynirvana** [comment]
8 bytes, it's sca I guess.
You can test with the new commit.

**Comment by pxia** [comment]
The alphabet of cookies is different from URL:
https://stackoverflow.com/questions/1969232/allowed-characters-in-cookies

Unlike URL parameters, there is no agreement how cookies data *must* be encoded and decoded, it's up to the server and client to interpret them differently. In your case, the `%` is not encoded data. It is the actual data. The `+` in `cna`  is also the actual data. 

However, in your case, the cookies seems needs to be *unquoted*, whereas the `cna` can be used without being unquoted. Again, the behavior is up to the clients.

**Comment by alkorsan** [comment]
@rosynirvana 
this is the debug of the last update eed9e2a

`
you-get --debug http://v.youku.com/v_show/id_XMjc4MjcwNjc4MA==.html
you-get: version 0.4.750, a tiny downloader that scrapes the web.
you-get: ['http://v.youku.com/v_show/id_XMjc4MjcwNjc4MA==.html']
Traceback (most recent call last):
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 1318, in do_open
    encode_chunked=req.has_header('Transfer-encoding'))
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/http/client.py", line 1239, in request
    self._send_request(method, url, body, headers, encode_chunked)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/http/client.py", line 1285, in _send_request
    self.endheaders(body, encode_chunked=encode_chunked)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/http/client.py", line 1234, in endheaders
    self._send_output(message_body, encode_chunked=encode_chunked)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/http/client.py", line 1026, in _send_output
    self.send(msg)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/http/client.py", line 964, in send
    self.connect()
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/http/client.py", line 936, in connect
    (self.host,self.port), self.timeout, self.source_address)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/socket.py", line 704, in create_connection
    for res in getaddrinfo(host, port, 0, SOCK_STREAM):
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/socket.py", line 743, in getaddrinfo
    for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
socket.gaierror: [Errno 7] No address associated with hostname

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/bin/you-get", line 11, in <module>
    load_entry_point('you-get==0.4.750', 'console_scripts', 'you-get')()
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/__main__.py", line 92, in main
    main(**kwargs)
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/common.py", line 1444, in main
    script_main('you-get', any_download, any_download_playlist, **kwargs)
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/common.py", line 1355, in script_main
    download_main(download, download_playlist, args, playlist, output_dir=output_dir, merge=merge, info_only=info_only, json_output=json_output, caption=caption)
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/common.py", line 1157, in download_main
    download(url, **kwargs)
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/common.py", line 1437, in any_download
    m.download(url, **kwargs)
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/extractor.py", line 42, in download_by_url
    self.prepare(**kwargs)
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/extractors/youku.py", line 191, in prepare
    data = youku_ups(self.vid)['data']
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/extractors/youku.py", line 39, in youku_ups
    url += '&utid=' + fetch_cna()
  File "/cygdrive/F/_bin/_bin/cygwin/home/LLED2/badrr/lib/python3.6/site-packages/you_get/extractors/youku.py", line 27, in fetch_cna
    req = urllib.request.urlopen(url)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 223, in urlopen
    return opener.open(url, data, timeout)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 526, in open
    response = self._open(req, data)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 544, in _open
    '_open', req)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 504, in _call_chain
    result = func(*args)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 1346, in http_open
    return self.do_open(http.client.HTTPConnection, req)
  File "/cygdrive/F/_bin/_bin/cygwin/lib/python3.6/urllib/request.py", line 1320, in do_open
    raise URLError(err)
urllib.error.URLError: <urlopen error [Errno 7] No address associated with hostname>
`

**Comment by rosynirvana** [comment]
How many times have you tried?

Maybe a network problem, I cannot reproduce
```
./you-get -di 'http://v.youku.com/v_show/id_XMjc4MjcwNjc4MA==.html'
[DEBUG] get_content: https://ups.youku.com/ups/get.json?vid=XMjc4MjcwNjc4MA==&ccode=0401&client_ip=192.168.1.1&utid=xoy4EYdwli8CAbZyXbW0QYSm&client_ts=1496424134
site:                优酷 (Youku)
title:               红糖CP〓我的世界-无人入眠〓①向流星许愿
streams:             # Available quality and codecs
    [ DEFAULT ] _________________________________
    - format:        hd3
      container:     flv
      video-profile: 1080P
      size:          359.3 MiB (376797251 bytes)
      m3u8_url:      http://pl-ali.youku.com/playlist/m3u8?vid=XMjc4MjcwNjc4MA%3D%3D&type=hd3&ups_client_netip=182.114.93.181&ups_ts=1496424137&utid=xoy4EYdwli8CAbZyXbW0QYSm&ccode=0401&psid=8beb2db7810a4b6ba44a6fc94a8df03c&ckey=&duration=1265&expire=18000&ups_key=31aca75734c024cde00f1db1cd5fcfa6
    # download-with: you-get --format=hd3 [URL]

    - format:        hd2
      container:     flv
      video-profile: 超清
      size:          186.8 MiB (195827087 bytes)
      m3u8_url:      http://pl-ali.youku.com/playlist/m3u8?vid=XMjc4MjcwNjc4MA%3D%3D&type=hd2&ups_client_netip=182.114.93.181&ups_ts=1496424137&utid=xoy4EYdwli8CAbZyXbW0QYSm&ccode=0401&psid=8beb2db7810a4b6ba44a6fc94a8df03c&ckey=&duration=1265&expire=18000&ups_key=31aca75734c024cde00f1db1cd5fcfa6
    # download-with: you-get --format=hd2 [URL]

    - format:        mp4
      container:     mp4
      video-profile: 高清
      size:          111.8 MiB (117259696 bytes)
      m3u8_url:      http://pl-ali.youku.com/playlist/m3u8?vid=XMjc4MjcwNjc4MA%3D%3D&type=mp4&ups_client_netip=182.114.93.181&ups_ts=1496424137&utid=xoy4EYdwli8CAbZyXbW0QYSm&ccode=0401&psid=8beb2db7810a4b6ba44a6fc94a8df03c&ckey=&duration=1265&expire=18000&ups_key=31aca75734c024cde00f1db1cd5fcfa6
    # download-with: you-get --format=mp4 [URL]

    - format:        flvhd
      container:     flv
      video-profile: 标清
      size:          57.6 MiB (60437246 bytes)
      m3u8_url:      http://pl-ali.youku.com/playlist/m3u8?vid=XMjc4MjcwNjc4MA%3D%3D&type=flv&ups_client_netip=182.114.93.181&ups_ts=1496424137&utid=xoy4EYdwli8CAbZyXbW0QYSm&ccode=0401&psid=8beb2db7810a4b6ba44a6fc94a8df03c&ckey=&duration=1265&expire=18000&ups_key=31aca75734c024cde00f1db1cd5fcfa6
    # download-with: you-get --format=flvhd [URL]
```

**Comment by alkorsan** [comment]
Hi,
I made a new install
pip3 install https://github.com/rosynirvana/you-get/archive/youku.zip
and I get the same strange error.
but if I remove eed9e2a and comeback to a0616b1 I get  客户端无权播放,201
thanks

**Comment by zywzyw1234567** [comment]
@rosynirvana 
是的，我也更新了youku.py，结果报错

debug结果如下：
```
C:\Users\zywzy>you-get --debug https://v.youku.com/v_show/id_XMjc5Nzg2MzQ4MA==.html?spm=a2hzp.8244740.0.0
you-get: version 0.4.750, a tiny downloader that scrapes the web.
you-get: ['https://v.youku.com/v_show/id_XMjc5Nzg2MzQ4MA==.html?spm=a2hzp.8244740.0.0']
Traceback (most recent call last):
  File "C:\Users\zywzy\AppData\Local\Programs\Python\Python36\Scripts\you-get-script.py", line 11, in <module>
    load_entry_point('you-get==0.4.750', 'console_scripts', 'you-get')()
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\__main__.py", line 92, in main
    main(**kwargs)
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\common.py", line 1444, in main
    script_main('you-get', any_download, any_download_playlist, **kwargs)
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\common.py", line 1355, in script_main
    download_main(download, download_playlist, args, playlist, output_dir=output_dir, merge=merge, info_only=info_only, json_output=json_output, caption=caption)
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\common.py", line 1157, in download_main
    download(url, **kwargs)
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\common.py", line 1437, in any_download
    m.download(url, **kwargs)
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\extractor.py", line 42, in download_by_url
    self.prepare(**kwargs)
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\extractors\youku.py", line 191, in prepare
    data = youku_ups(self.vid)['data']
  File "c:\users\zywzy\appdata\local\programs\python\python36\lib\site-packages\you_get\extractors\youku.py", line 39, in youku_ups
    url += '&utid=' + fetch_cna()
TypeError: must be str, not NoneType
```

**Comment by pxia** [comment]
```
>>> import urllib.request
>>> url = 'http://gm.mmstat.com/yt/ykcomment.play.commentInit?cna='
>>> headers = urllib.request.urlopen(url).getheaders()
>>> print(headers)
[('Server', 'Tengine'), ('Date', 'Fri, 02 Jun 2017 20:24:09 GMT'), ('Content-Type', 'image/gif'), ('Content-Length', '43'), ('Connection', 'close'), ('P3P', 'CP="NOI DSP COR CURa ADMa DEVa PSAa PSDa OUR IND UNI PUR NAV"'), ('Set-Cookie', 'sca=80810f82; path=/; domain=.mmstat.com'), ('Expires', 'Thu, 01 Jan 1970 00:00:01 GMT'), ('Cache-Control', 'no-cache'), ('Pragma', 'no-cache')]
```

所以因为这个`('Set-Cookie', 'sca=80810f82; path=/; domain=.mmstat.com')`，是不是要改一下这里

```
if name == 'cna':
    return quote_cna(value)
```
改成`if name == 'sca'`?



另外`url = 'http://gm.mmstat.com/yt/ykcomment.play.commentInit?cna='`后面的cna没有值，可以去掉吗？

**Comment by rosynirvana** [comment]
要的是cna这个cookie改成sca当然不行啊

**Comment by rosynirvana** [comment]
@alkorsan reinstalling won't help you if you runned into something like a network problem

**Comment by pxia** [comment]
你看我上面那个response里，根本没有`cna`

**Comment by StLym** [comment]
有没有大佬解决了这个问题，换了几个youku.py也没用啊

**Comment by soimort** [comment]
Closed via #2015.
